pkg/db/subscription.go

```diff
@@ -0,0 +1,92 @@
+package db
+
+import (
+	"fmt"
+	"time"
+)
+
+type PollableDBQuery[ValueType any] func(lastSeenID int64, numRows int32) (results []ValueType, lastID int64, err error)
+
+// Poll whenever notified, or at an interval if not notified
+type PollingOptions struct {
+	Interval time.Duration
+	Notifier <-chan bool
+	NumRows  int32
+}
+
+type DBSubscription[ValueType any] struct {
+	lastSeenID int64
+	options    PollingOptions
+	query      PollableDBQuery[ValueType]
+	updates    chan<- []ValueType
+}
+
+func NewDBSubscription[ValueType any](
+	query PollableDBQuery[ValueType],
+	lastSeenID int64,
+	options PollingOptions,
+) *DBSubscription[ValueType] {
+	return &DBSubscription[ValueType]{
+		lastSeenID: lastSeenID,
+		options:    options,
+		query:      query,
+		updates:    nil,
+	}
+}
+
+func (s *DBSubscription[ValueType]) Start() (<-chan []ValueType, error) {
+	var err error
+	if s.updates != nil {
+		return nil, fmt.Errorf("Already started")
+	}
+	updates := make(chan []ValueType)
+	s.updates = updates
+
+	go func() {
+		if err := s.poll(); err != nil {
+			panic("TODO(rich)")
+		}
+
+		for {
+			select {
+			case <-s.options.Notifier:
+				if err = s.poll(); err != nil {
+					panic("TODO(rich)")
+				}
+			case <-time.After(s.options.Interval):
```

Comment:
`time.After` is a Go feature that's a bit of a trap. `time.Ticker` is [weirdly more performant](https://github.com/ChainSafe/gossamer/issues/1623) and doesn't have [known memory leaks](https://100go.co/#timeafter-and-memory-leaks-76).